Extract sensor manager

scytacki opened the pull request:
This pulls out the SensorConnectorInterface bits into a SensorConnectorManager class with an ISensorManager interface.  It also includes a FakeSensorManager for testing without sensors.

It could be improved more, but it might be worth merging even before the improvements.  To merge we'd need to resolve some merge conflicts.

The further improvements are:
- abstraction of the SensorConfiguration, currently this uses the ISensorConfig from sensor-connector-interface. ISensorConfig includes lots of fields that are are not needed.
- abstraction of ISensorConfigColumnInfo this is used in several places. This is bad because it includes properties specific to the SensorConnector API. It is also bad because it includes a data field which if used would break the onSensorData abstraction.

kswenson commented on src/components/app.tsx:
```diff
@@ -213,95 +207,119 @@ export class App extends React.Component<AppProps, AppState> {
-        this.setState({ sensorSlots: this.state.sensorSlots });
-    }
-    
     sensorHasData():boolean {
```
Consider `hasSensorData()`
- `sensorHasData()` reads to me like a question about a single sensor, while `hasSensorData()` reads more like a question about whether sensor data is available. (May just be me.)
- I generally try to follow the convention that boolean accessors should start with the question, e.g. `is...()`, `has...()`.

scytacki commented on src/components/app.tsx:
```diff
@@ -119,25 +118,23 @@ export class App extends React.Component<AppProps, AppState> {
+        let listeners = this.props.sensorManager.listeners;
+        listeners.onSensorConnect = this.onSensorConnect;
+        listeners.onSensorData = this.onSensorData;
+        listeners.onSensorStatus = this.onSensorStatus;
```
@kswenson do you have a preference between the 1st and 3rd observer pattern options described here:   https://github.com/millermedeiros/js-signals/wiki/Comparison-between-different-Observer-Pattern-implementations 
The 2nd option doesn't seem good because I'm not going to change everything in the project to use a single dispatcher. I like the type checking of the 3rd option, but it is not standard.

kswenson replied:
I was imagining something more like the first option, in part because it seems like a less radical change from what we have. For instance, the client could call:
```
sensorManager.addListener('onSensorConnect', this.onSensorConnect);
```
and later:
```
sensorManager.removeListener('onSensorConnect', this.onSensorConnect);
```
and the SensorManager itself would have to change relatively little.

scytacki commented:
@kswenson so I think I've addressed all of the issues you raised with the exception of the addListener or observer API. See my comment above for a question about that.
